token/lib/token.js

```diff
@@ -100,6 +110,148 @@ class Token {
+   */
+  async unpause(networkId) {
+    const contract = await this.contract(networkId)
+    const sender = this.defaultAccount(networkId)
+
+    // Pre-contract call validations.
+    const paused = await contract.methods.paused().call()
+    if (!paused) {
+      throw new Error('Token is already unpaused')
+    }
+    const tokenOwner = await contract.methods.owner().call()
+    if (tokenOwner.toLowerCase() != sender.toLowerCase()) {
+      throw new Error(`Sender ${sender} is not owner of token contract (${tokenOwner})`)
+    }
+
+    const transaction = contract.methods.unpause()
+    await this.sendTransaction(networkId, transaction, { from: sender })
+    if (await contract.methods.paused().call() !== false) {
+      throw new Error('Token should be unpaused but is not')
+    }
+  }
+
+  // TODO: refactor into separate base class, to support other contracts such
+  // as the marketplace contract
+  /**
+   * Sends an Ethereum transaction.
+   * @param {string} networkId - Ethereum network ID.
+   * @param {transaction} transaction - These are returned by contract.methods.MyMethod()
+   * @param {Object} opts - Options to be sent along with the transaction.
+   * @returns {Object} - Transaction receipt.
+   */
+  async sendTransaction(networkId, transaction, opts = {}) {
+    // TODO: support multisig wallets
+
+    const web3 = this.web3(networkId)
+
+    let transactionHash
+    if (!opts.gas) {
+      opts.gas = await transaction.estimateGas({ from: opts.from })
+      this.vlog('estimated gas:', opts.gas)
+    }
+
+    // Send the transaction and grab the transaction hash when it's available.
+    this.vlog('sending transaction')
+    transaction.send(opts)
+      .on('transactionHash', (hash) => {
+        transactionHash = hash
+        this.vlog('transaction hash:', transactionHash)
+      })
+
+    // Poll for the transaction receipt, with an exponential backoff. This works
+    // around some strange interactions between web3.js and some web3 providers.
+    // For example, this issue sometimes prevents transaction receipts from
+    // being returned when simply calling
+    // `await contract.methods.MyMethod(...).send(...)`:
+    //
+    // https://github.com/INFURA/infura/issues/95
+    const maxSleep = 15 * 60 * 1000 // 15 minutes, in milliseconds
+    let totalSleep = 0
+    let sleepTime = 1000
+    while (totalSleep <= maxSleep) {
+      await sleep(sleepTime)
+
+      if (transactionHash) {
+        const receipt = await web3.eth.getTransactionReceipt(transactionHash)
+        if (receipt) {
+          this.vlog('got transaction receipt', receipt)
+          if (receipt.status) {
+            this.vlog('transaction successful')
+            return receipt
+          } else {
+            throw new Error('transaction failed')
+          }
+        } else {
+          this.vlog('waiting for transaction receipt')
+        }
+      } else {
+        this.vlog('waiting for transaction hash')
```

Comment:
Nit: maybe log how many sec the next check will be performed. So that the operator running the script does not think it is hanging.

Reply:
Good point. I've removed this log message and added another at the beginning of the `while`.